File: 2015/cpp/day01.cpp
#include <iostream>
#include <vector>
#include <string>
#include <fstream>
#include <algorithm>
using namespace std;


vector<string> readFileString() {
    
    ifstream file("../input/input01.txt");
    string str;
    vector<string> lineas;

    while (getline(file, str)) {
        if(str.size() > 0){
            lineas.push_back(str+'\0');
        }
    }
    file.close();

    return lineas;
}

int part2(vector<string> lineas){
    int piso = 0;
    for(int n = 0; n < lineas[0].length(); n++){
        if (lineas[0][n] == '(') 
            piso++;
        else
            piso--;
        if (piso == -1)
            return n+1;
    }

    return 0;
}


int part1(vector<string> lineas){
    int up = count(lineas[0].begin(), lineas[0].end(), '(');
    int down = count(lineas[0].begin(), lineas[0].end(), ')');
    
    return up-down;
}

int main () 
{   
    vector<string> lineas = readFileString();
    cout << "Part 1, Solución: " << part1(lineas) << endl;
    cout << "Part 2, Solución: " << part2(lineas) << endl;
    return 0;
}

File: 2015/java/Day01.java
import java.util.*;
import java.util.stream.*;
import java.io.*;
import java.nio.file.*;

public class Day01 {
    private List<String> lineas = new ArrayList<>();

    Day01() {
        readInput();
    }

    public int Parte1() {
        int count1 = (int) lineas.get(0).codePoints().filter(ch -> ch == '(').count();
        int count2 = (int) lineas.get(0).codePoints().filter(ch -> ch == ')').count();

        return count1-count2;
    }

    public int Parte2() {
        int level = 0;
        for (int i=0; i<lineas.get(0).length(); i++){
            if (lineas.get(0).charAt(i) =='(') {
                level++;
            } else {
                level--;
            }

            if (level == -1) {
                level = i + 1;
                break;
            }
        }
        return level;
    }

    private void readInput() {
        String fileName = "../input/input01.txt";

        try {
            Stream<String> stream = Files.lines(Paths.get(fileName));
            lineas = stream.collect(Collectors.toList());
        } catch (Exception e) {
            System.out.println(e);
        }
    }


   public static void main(String[] args) {
      Day01 day01 = new Day01();
      System.out.println("Solución: " + day01.Parte1());
      System.out.println("Solución: " + day01.Parte2());
   }
}